docs/architecture/adr-038-state-listening.md

```diff
@@ -0,0 +1,612 @@
+to wrap the returned `KVStore` with a `listenkv.Store` if listening is turned on for that `Store`.
+
+```go
+func (rs *Store) GetKVStore(key types.StoreKey) types.KVStore {
+	store := rs.stores[key].(types.KVStore)
+
+	if rs.TracingEnabled() {
+		store = tracekv.NewStore(store, rs.traceWriter, rs.traceContext)
+	}
+	if rs.ListeningEnabled(key) {
+		store = listenkv.NewStore(key, store, rs.listeners[key])
+	}
+
+	return store
+}
+```
+
+We will also adjust the `cachemulti` constructor methods and the `rootmulti` `CacheMultiStore` method to forward the listeners
+to and enable listening in the cache layer.
+
+```go
+func (rs *Store) CacheMultiStore() types.CacheMultiStore {
+	stores := make(map[types.StoreKey]types.CacheWrapper)
+	for k, v := range rs.stores {
+		stores[k] = v
+	}
+	return cachemulti.NewStore(rs.db, stores, rs.keysByName, rs.traceWriter, rs.traceContext, rs.listeners)
+}
+```
+
+### Exposing the data
+
+We will introduce a new `StreamingService` interface for exposing `WriteListener` data streams to external consumers.
+
+```go
+// Hook interface used to hook into the ABCI message processing of the BaseApp
+type Hook interface {
+	ListenBeginBlock(ctx sdk.Context, req abci.RequestBeginBlock, res abci.ResponseBeginBlock) // update the streaming service with the latest BeginBlock messages
+	ListenEndBlock(ctx sdk.Context, req abci.RequestEndBlock, res abci.ResponseEndBlock) // update the steaming service with the latest EndBlock messages
+	ListenDeliverTx(ctx sdk.Context, req abci.RequestDeliverTx, res abci.ResponseDeliverTx) // update the steaming service with the latest DeliverTx messages
+}
+
+// StreamingService interface for registering WriteListeners with the BaseApp and updating the service with the ABCI messages using the hooks
+type StreamingService interface {
+	Stream(wg *sync.WaitGroup, quitChan <-chan struct{}) // streaming service loop, awaits kv pairs and writes them to some destination stream or file
+	Listeners() map[sdk.StoreKey][]storeTypes.WriteListener // returns the streaming service's listeners for the BaseApp to register
+	Hook
+}
+```
+
+#### Writing state changes to files
+
+We will introduce an implementation of `StreamingService` which writes state changes out to files as length-prefixed protobuf encoded `StoreKVPair`s.
+This service uses the same `StoreKVPairWriteListener` for every KVStore, writing all the KV pairs from every KVStore
+out to the same files, relying on the `StoreKey` field in the `StoreKVPair` protobuf message to later distinguish the source for each pair.
+
+The file naming schema is as such:
+* After every `BeginBlock` request a new file is created with the name `block-{N}-begin`, where N is the block number. All
+subsequent state changes are written out to this file until the first `DeliverTx` request is received. At the head of these files,
+  the length-prefixed protobuf encoded `BeginBlock` request is written, and the response is written at the tail.
```

Comment:
How do you separate the response from the state changes when parsing?

Reply:
In this current format the only way to distinguish the response is that it is the final length-prefix encoded protobuf entry in the file. This can work if we buffer the bytes of each file fully into memory, segmenting it into separate protobuf objects by the length-prefixes, whereby we can recognize which object is the last before we attempt to decode it.

This does feel kinda janky! But I think it could be feasible since each file is just one ABCI stage of a Tendermint block, we already need all these bytes to fit in memory since we have to hold it all in memory before we flush it out to the file in the first place.  Do you think it'd be better to add something like a total-length prefix to the beginning of the state changes? E.g. a variable length prefix in front of the first length-prefixed state change that encodes the entire length of all the prefixed state changes.